src/main/java/bio/terra/service/upgrade/flight/BackfillRowMetadataTablesStep.java

```diff
@@ -0,0 +1,116 @@
+package bio.terra.service.upgrade.flight;
+
+import static bio.terra.service.tabulardata.google.BigQueryPdao.prefixName;
+
+import bio.terra.model.EnumerateDatasetModel;
+import bio.terra.model.EnumerateSortByParam;
+import bio.terra.model.SqlSortDirection;
+import bio.terra.model.UpgradeModel;
+import bio.terra.service.dataset.Dataset;
+import bio.terra.service.dataset.DatasetService;
+import bio.terra.service.iam.AuthenticatedUserRequest;
+import bio.terra.service.iam.IamResourceType;
+import bio.terra.service.iam.IamService;
+import bio.terra.service.tabulardata.google.BigQueryPdao;
+import bio.terra.service.tabulardata.google.BigQueryProject;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.exception.RetryException;
+import java.util.Set;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BackfillRowMetadataTablesStep implements Step {
+  private static final Logger logger = LoggerFactory.getLogger(BackfillRowMetadataTablesStep.class);
+  private final UpgradeModel request;
+  private final DatasetService datasetService;
+  private final IamService iamService;
+  private final AuthenticatedUserRequest userReq;
+  private final BigQueryPdao bigQueryPdao;
+
+  BackfillRowMetadataTablesStep(
+      UpgradeModel request,
+      DatasetService datasetService,
+      IamService iamService,
+      BigQueryPdao bigQueryPdao,
+      AuthenticatedUserRequest userReq) {
+    this.request = request;
+    this.datasetService = datasetService;
+    this.iamService = iamService;
+    this.userReq = userReq;
+    this.bigQueryPdao = bigQueryPdao;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    // enumerate datasets
+    // Appears the max in an environment is 188 in dev
+    int limit = 1000;
+    int skip = 0;
+    int totalDatasetCount = 0;
+    Set<UUID> resources =
+        iamService.listAuthorizedResources(userReq, IamResourceType.DATASET).keySet();
+    EnumerateDatasetModel enumerateDatasetModel =
+        datasetService.enumerate(
+            skip, limit, EnumerateSortByParam.NAME, SqlSortDirection.ASC, null, null, resources);
```

Comment:
I think it's fine for now given that we know exactly how many datasets there are but in the future, you could do a while loop that keeps paging until the result list is empty

Reply:
Agreed